### PR TITLE
Configure Test Panel for Initializing Diffraction Calibration

### PR DESCRIPTION
**Pull request by darshdinger**
For EW1815

### CODE REVIEW

File: src/snapred/backend/service/InitializeCalibrationService.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            raise ValueError("List is empty")
+
+        else:
+            # list to store states
+            states = []
+
+            for run in runs:
+                reductionIngredients = self.dataFactory.getReductionIngredients(run.runNumber)
+                ipts = reductionIngredients.runConfig.IPTS
+                ipts + "shared/lite/SNAP_{}.lite.nxs.h5".format(reductionIngredients.runConfig.runNumber)
```
Comment (walshmm) [inline src/snapred/backend/service/InitializeCalibrationService.py:55]: I assume the formatter removed the assignment of this concatenation to some variable?
Also if you just need IPTS consider exposing [this method](https://github.com/neutrons/SNAPRed/blob/6f605996d68e237361370a2b602f59f436fde112/src/snapred/backend/data/LocalDataService.py#L157) to the datafactoryservice

File: src/snapred/ui/presenter/LogTablePresenter.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def updateButton(self, text):
     def handle_button_clicked(self):
         self.view.button.setEnabled(False)
 
-        request = SNAPRequest(path="/reduction", payload=RunConfig(runNumber="48741").json())
+        request = SNAPRequest(path="/reduction", payload=RunConfig(runNumber="45158").json())
```
Comment (walshmm) [inline src/snapred/ui/presenter/LogTablePresenter.py:78]: Is there a reason for the changed runNumber?

Comment (darshdinger) [reply src/snapred/ui/presenter/LogTablePresenter.py:78]: I was running some manual testing!

File: src/snapred/ui/presenter/InitializeCalibrationPresenter.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    def handlePixelGroupingResult(self, response: SNAPResponse):
+        if response.responseCode == 200:
+            self._labelView("Ready to Calibrate!")
+
+    def _spawnStateCreationWorkflow(self):
+        from snapred.ui.widget.WorkflowNode import continueWorkflow, finalizeWorkflow, startWorkflow
+
+        promptView = PromptUserforCalibrationInputView()
+
+        def push_data_to_interface_controller():
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeCalibrationPresenter.py:94]: Will do!

File: src/snapred/ui/presenter/LogTablePresenter.py (hunk 2 above)
Comment (darshdinger) [reply src/snapred/ui/presenter/LogTablePresenter.py:78]: I will revert the changes!

File: src/snapred/ui/presenter/InitializeCalibrationPresenter.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    def _labelView(self, text: str):
+        win = QWidget()
+        vbox = QVBoxLayout()
+        label = QLabel(text)
+        label.setAlignment(Qt.AlignCenter)
+        vbox.addWidget(label)
+        vbox.addStretch()
+        win.setLayout(vbox)
+        self.view.layout().addWidget(win)
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeCalibrationPresenter.py:29]: Okay, I can do that!

File: src/snapred/backend/service/ServiceFactory.py [hunk 5]
```diff
@@ -26,14 +27,15 @@ class ServiceFactory:
...
-        self.serviceDirectory.registerService(ApiService)
-        self.serviceDirectory.registerService(FitMultiplePeaksService)
+        self.serviceDirectory.registerService(ConfigLookupService())
+        self.serviceDirectory.registerService(ReductionService())
+        self.serviceDirectory.registerService(StateIdLookupService())
+        self.serviceDirectory.registerService(CalibrationService())
+        self.serviceDirectory.registerService(CrystallographicInfoService())
+        self.serviceDirectory.registerService(CalibrantSampleService())
+        self.serviceDirectory.registerService(ApiService())
+        self.serviceDirectory.registerService(FitMultiplePeaksService())
```
Comment (walshmm) [inline src/snapred/backend/service/ServiceFactory.py:37]: These shouldnt be instantiated.  Pete changed it to lazily initialize.

File: src/snapred/ui/presenter/InitializeCalibrationPresenter.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        self.worker_pool.submitWorker(self.worker)
+
+    def handleStateCheckResult(self, response: SNAPResponse):
+        if response.responseCode != 200:
+            self._spawnStateCreationWorkflow()
+            return
+        else:
+            pass
+
+        groupingFile = str(StateConfig.focusGroups.definition)  # This is incorrect, how do I find the groupingFile??
```
Comment (walshmm) [inline src/snapred/ui/presenter/InitializeCalibrationPresenter.py:73]: I think this needs to be populated with a real value.  You may need to write up a separate method that looks up the focusGroups and feeds them to calculatePixelGroupingParameters.  You  can look them up by exposing[ this method](https://github.com/neutrons/SNAPRed/blob/next/src/snapred/backend/data/LocalDataService.py#L161) to the DataFactoryService.  So then you can just foreach over the groups.

Not saying you need to do something about this -> but, Realistically "/calibration/calculatePixelGroupingParameters" shouldnt be user facing as we wouldnt expect them to provide custom group files.

File: src/snapred/ui/view/InitializeCalibrationCheckView.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+from PyQt5.QtCore import *
+from PyQt5.QtGui import *
+from PyQt5.QtWidgets import *
```
Comment (walshmm) [inline src/snapred/ui/view/InitializeCalibrationCheckView.py:3]: Change these from wildcard imports to specific imports, I would have thought precommit would have caught that.

File: src/snapred/backend/recipe/algorithm/IngestCrystallographicInfoAlgorithm.py
Comment (darshdinger) [inline src/snapred/backend/recipe/algorithm/IngestCrystallographicInfoAlgorithm.py:1]: This is meant for another PR. Git hub did a weird commit thing.

File: src/snapred/backend/service/ServiceFactory.py (hunk 5 above)
Comment (darshdinger) [reply src/snapred/backend/service/ServiceFactory.py:37]: Okay, fixed it.

File: src/snapred/ui/presenter/InitializeCalibrationPresenter.py (hunk 6 above)
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeCalibrationPresenter.py]: Please reference edd5226 for fixes. Hopefully this works?

File: tests/unit/ui/presenter/test_InitializeCalibrationPresenter.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    view = Mock()
+    view.beginFlowButton = Mock()
+    view.beginFlowButton.setEnabled = Mock()
+    view.getRunNumber = Mock(return_value=12345)
+    view.layout = Mock()
+    view.layout().addWidget = Mock()
+    return view
+
+
+def test_labelView(mock_View, qtbot):
```
Comment (walshmm) [inline tests/unit/ui/presenter/test_InitializeCalibrationPresenter.py:21]: you can add a comment to the end of the line in the fashion of 
```  #noqa: ARG001```
and precommit will let it pass.

File: src/snapred/backend/service/CalibrationService.py [hunk 9]
```diff
@@ -120,6 +125,27 @@ def calculatePixelGroupingParameters(self, runs: List[RunConfig], groupingFile:
...
+        stateID, _ = self.localDataService._generateStateId(runId)
+        calibrationStatePath = self.localDataService._constructCalibrationPath(stateID)
+
+        if os.path.exists(calibrationStatePath):
+            recordPath: str = calibrationStatePath + "{}/v{}/CalibrationRecord.json".format(runId, version)
+            if os.path.exists(recordPath):
+                return True
+            else:
+                return False
+        else:
```
Comment (walshmm) [inline src/snapred/backend/service/CalibrationService.py:139]: most of this should live in the localdataservice as you are doing filesystem level things, a service should be agnostic of the filesystem implementation

File: src/snapred/backend/service/CalibrationService.py [hunk 10]
```diff
@@ -120,6 +125,27 @@ def calculatePixelGroupingParameters(self, runs: List[RunConfig], groupingFile:
...
+            return False
+
+    @FromString
+    def checkDataExists(self, run: RunConfig):
```
Comment (walshmm) [inline src/snapred/backend/service/CalibrationService.py:143]: what is the purpose of this method?

File: src/snapred/backend/service/CalibrationService.py (hunk 9 above)
Comment (walshmm) [reply src/snapred/backend/service/CalibrationService.py:139]: (members prepended with an underbar `_` are considered private and shouldnt be called outside their class usually)

Comment (walshmm) [reply src/snapred/backend/service/CalibrationService.py:139]: you could in theory replace this code with a None check on the return of [this](https://github.com/neutrons/SNAPRed/blob/next/src/snapred/backend/data/LocalDataService.py#L464)
Though you would still need the preliminary check to see if the state folder even exists.

Comment (darshdinger) [reply src/snapred/backend/service/CalibrationService.py]: Should be fixed.... hopefully.

**Review by walshmm — APPROVED** [review]

**Pull request merged**